### DISCUSSION TITLE
nil pointer error in device-sdk-go when I run device-mqtt-go

### DISCUSSION THREAD

**Discussion by cglin7** [edgexfoundry/community#166] [Q&A] [answered]
![1692080951950](https://github.com/edgexfoundry/community/assets/31946087/7bfb1921-e588-4549-90fc-f17d78c42b6f)

seems like some "Client" communicate with ”core-metadata“ wasn't find in DI container，so I add these code to device-sdk-go，and device-mqtt-go service run well，without these code，the service throws nil pointer error。But，I don‘t think adding code to device-sdk-go is the right way，how can I solve this nil pointer error？

**Comment by cloudxxx8** [comment]
All the core-metadata clients are in the DI.  You don't need to modify any code.
https://github.com/edgexfoundry/device-sdk-go/blob/2bcb30427b801ff9ed3b105e71011a55aed1d51b/pkg/service/service.go#L142C13-L142C32

See the usage example:
https://github.com/edgexfoundry/device-sdk-go/blob/2bcb30427b801ff9ed3b105e71011a55aed1d51b/pkg/service/manageddevices.go#L38C5-L38C5

**Reply by cglin7** [reply]
ok, I see, thank you

**Comment by bnevis-i** [comment]
The location where the clients are SUPPOSED to be pre-created in the DI is:

https://github.com/edgexfoundry/go-mod-bootstrap/blob/main/bootstrap/handlers/clients.go

**Reply by cglin7** [reply]
I see, the client config was missing in file ”configuration.yaml“ which is in folder cmd, I added the client config and it runs well，thank you

**Comment by lenny-goodell** [comment, marked as answer]
@cglin7 , I assume you are running Device MQTT from command line in hybrid mode (other services running in docker). 
If so, in 3.0 you need to provide the command line flags to let the service know where to get the common config (contains client config).
Please see this documentation for details: https://docs.edgexfoundry.org/3.0/getting-started/Ch-GettingStartedHybrid/#run-the-service-code-natively

Also note that the handling of error cause by missing common configuration have been improved for the upcoming 3.1 release on the main branch.

**Reply by cglin7** [reply]
Thanks for replying, I do run these services in hybrid mode，the device-mqtt-go service runs by local code, others services run by docker. The client config was missing so I added it and now it runs well

**Reply by lenny-goodell** [reply]
Adding that is not required if you follow the instructions in  https://docs.edgexfoundry.org/3.0/getting-started/Ch-GettingStartedHybrid/#run-the-service-code-natively. ie. use the `-cp -d- o` flags so the common config is loaded (has the clients) and runs service in Dev (hybrid) mode.
